Kernel/Syscalls/execve.cpp

```diff
@@ -215,30 +215,32 @@ static ErrorOr<FlatPtr> get_load_offset(const ElfW(Ehdr) & main_program_header,
     if (main_program_header.e_type != ET_EXEC)
         return EINVAL;
 
+    // Note: If we just need to map the main program and no interpreter (so we map a static binary),
+    // we can just map it anywhere the executable wants to be mapped, because we can't ever conflict.
+    if (!interpreter_description)
+        return 0;
```

Comment:
This kind of implies though that this is a non-PIE/PIC (I can never tell which is which) binary, right? As far as I know there are position-independent static executables that should receive the proper amount of randomization. Or are these the weird self-relocating binaries that we talked about at some point in the past?

Reply:
Right, this does not support other types of executables, which are PIE/PIC/whatever. IIRC one type is using relocations so you need to fix its offset value to ensure it can jump to the correct positions with the actual code.

Reply:
Does this mean that `Loader.so` will no longer be mapped at a random address? It's a static PIE that self-relocates.

I'd much prefer to move the self-relocation code from `LibELF/Relocation.cpp` to a static crt object. That would make `LibMainStatic` redundant and would also help in @sin-ack's zig port.

Reply:
> Does this mean that Loader.so will no longer be mapped at a random address?

I don't think so. Loader is usually used as an interpreter, so it will hit the `interpreter_description` code path (see `Process::load`, which directly maps that binary, and doesn't go for another round of `get_load_offset`).

Reply:
That said, this is evidence that loading the Loader at randomized addresses works for all the wrong reasons. We should really look into a proper heuristic for whether a binary can be relocated instead of "this has an interpreter" and "this doesn't have an interpreter". I'll also try to find some time to refactor this function in particular to make it less dependent on the "main program"/"interpreter" distinction.